### PR TITLE
Gen4: groupby support

### PR DESCRIPTION
**Pull request by harshit-gangal**
## Description

This PR adds support for grouping and aggregations.
Gen4 still needs to handle grouping with distinct.

## Related Issue(s)

#7280 

## Checklist
- [x] Tests were added or are not required
- [x] Documentation was added or is not required

### CODE REVIEW

File: go/vt/vtgate/planbuilder/selectGen4.go [hunk 1]
```diff
@@ -218,22 +354,23 @@ func needsWeightString(tbl semantics.TableInfo, colName *sqlparser.ColName) bool
...
+		newLeft, _, err := planOrderBy(qp, orderExprs, plan.Left, semTable)
 		if err != nil {
-			return nil, err
+			return nil, false, err
 		}
 		plan.Left = newLeft
-		return plan, nil
+		return plan, false, nil
 	}
+	ms, colAdded, err := createMemorySortPlan(plan, orderExprs, semTable)
```
Comment (systay) [inline go/vt/vtgate/planbuilder/selectGen4.go:366]: we could just return here, no need to check `err`

File: go/vt/vtgate/planbuilder/selectGen4.go [hunk 2]
```diff
@@ -245,45 +382,65 @@ func planOrderByForJoin(qp *queryProjection, orderExprs []orderBy, plan *joinGen
 	}
 
 	for _, order := range orderExprs {
-		expr := order.inner.Expr
-		offset, err := wrapExprAndPush(expr, plan, semTable)
-		if err != nil {
-			return nil, err
+		offset, woffset, found := findExprInOrderedAggr(plan, order)
+		if !found {
+			return nil, vterrors.New(vtrpcpb.Code_INTERNAL, "expected to find this expression")
```
Comment (systay) [inline go/vt/vtgate/planbuilder/selectGen4.go:387]: more better error message please

**Review by systay — APPROVED** [review]

**Review by frouioui — APPROVED** [review]

File: go/vt/vtgate/planbuilder/selectGen4.go (hunk 1 above)
Comment (frouioui) [reply go/vt/vtgate/planbuilder/selectGen4.go]: Resolved via 03aa03adb0

File: go/vt/vtgate/planbuilder/selectGen4.go (hunk 2 above)
Comment (frouioui) [reply go/vt/vtgate/planbuilder/selectGen4.go:383]: How does that sound?
```go
return nil, vterrors.New(vtrpcpb.Code_INTERNAL, "expected to find order by expression in orderedAggregate")
```

**Review by GuptaManan100 — APPROVED** [review]
LGTM

**Comment by frouioui** [comment]
This pull request is related to #7280

File: go/vt/vtgate/planbuilder/abstract/queryprojection.go [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+	}
+
+	// QueryProjection contains the information about the projections, group by and order by expressions used to do horizon planning.
+	QueryProjection struct {
+		// If you change the contents here, please update the toString() method
+		SelectExprs        []SelectExpr
+		HasAggr            bool
+		GroupByExprs       []GroupBy
+		OrderExprs         []OrderBy
+		CanPushDownSorting bool
```
Comment (harshit-gangal) [inline go/vt/vtgate/planbuilder/abstract/queryprojection.go:43]: hasAggr and CanPushDownSorting should be added in toString method and a test can be added as well.

**Pull request merged**